docs/README.md

```diff
@@ -639,6 +641,18 @@ This project layout is based directly on https://blog.gruntwork.io/how-to-manage
 
 See also [Resource locations](#resource-locations) for a table referencing these terraform envs.
 
+# Bakeries: querying job logs
+
+To query logs for recipe runs, see the API docs at the `/docs#/logs` route of the deployment
+(https://api.pangeo-forge.org/docs#/logs for the production deployment).
+These routes are protected, due to the risk of leaking secrets in the logs. The
+`PANGEO_FORGE_API_KEY` is available to admins via in the deployments secrets config:
```

Comment:
@cisaacstern, i'm curious... does this mean that we won't be exposing the logs on the front-end in the near future? Is the idea that if something is broken, as an admin I can query the API to figure out what's going on but any other user won't be able to see these logs?

Reply:
@andersy005, based on @yuvipanda's comment https://github.com/pangeo-forge/pangeo-forge-orchestrator/issues/145#issuecomment-1261537417, I thought the careful thing would be to keep these protected to start, until we've had a chance to think through if/what to do with them.

> Is the idea that if something is broken, as an admin I can query the API to figure out what's going on but any other user won't be able to see these logs?

Initially, I was imagining that this might be the approach. But to be honest, I've been doing this for a while myself now (albeit without the help of an Orchestrator API route)... and it can be toilsome.

So perhaps the best thing is to just make these public to start with?

A third option would be leave the raw logs served here protected, but add another layer/route that applies some type of filtering to them, and the filtered/sanitized result could be passed to the frontend?

Reply:
@andersy005, for context, this is an example of the text that this route would return:

> https://gist.github.com/cisaacstern/2a79707feaf27c5c0a2d4d93e5738fe5

So entirely aside from leaking secrets, there's the question of how to format this into something that a frontend user would actually find useful (99% of it is apache beam boilerplate)...

Reply:
Basically only [the last ~4 lines](https://gist.github.com/cisaacstern/2a79707feaf27c5c0a2d4d93e5738fe5#file-example-logs-txt-L665-L668) are relevant to actually debugging the recipe.

Reply:
One more thought: this route doesn't offer any time of "follow"/"tail" functionality, and I'm not sure that it would perform especially well in a "stream logs to the frontend" capacity. Though conceivably we could do some string formatting to provide just a short error trace from the end of these logs, and display it on the frontend if `recipe_run.conclusion = failed"?

Reply:
> Yes, I was imagining that this might be the approach initially. 

I see... Perhaps focusing on de-risking our implementation is the right approach for a first try.  Once we are ready to make access to the logs fully public, we can revisit this. 

> One more thought: this route doesn't offer any time of "follow"/"tail" functionality, and I'm not sure that it would perform especially well in a "stream logs to the frontend" capacity.

Are performance concerns the main reason for not supporting "stream-like" events? Is this feature not available via `gcloud logging read`? I was wondering if this could be helpful


https://amittallapragada.github.io/docker/fastapi/python/2020/12/23/server-side-events.html

???

> So entirely aside from leaking secrets, there's the question of how to format this into something that a frontend user would actually find useful (99% of it is apache beam boilerplate)...

Debugging these long tracebacks will probably be a challenge. Nonetheless, I think providing easy access to the logs would be a great feature. Can these logs be filtered and structured (by another service/application) so they can be easily consumed by the front-end?

Reply:
> Can these logs be filtered and structured ... so they can be easily consumed by the front-end?

@andersy005, in the latest commits I've added a public `/logs/trace` route which truncates the last portion of the traceback (for failed runs only) and serves it over a public route for the frontend to consume. Your feedback pushed me to figure out some way to make this useful _now_, and this is what I came up with. I totally agree that this is an essential feature to lighten the load on maintainers (I've been there!) and empower recipe contributors to be more self-sufficient.

There will no doubt be edge cases to deal with, but hopefully this feature will useful in its current form, at least in certain circumstances. Once this PR is merged, I'll make a recommendation for next steps for the logging feature in a new issue.